File: 46. Permutations.swift
class Solution {

    // Arbitary init value
    var nums: Array<Int> = []
    var result: Array<Array<Int>> = []
    
    private func backtrackingDFS(_ value: Int,_ set: Set<Int>,_ array: Array<Int>){
                
        if array.count == nums.count {
            result.append(array)
            return
        }
        
        var newSet = set
        newSet.insert(value)
        
        for num in nums {
            if !newSet.contains(num){
                var newArray = array
                newArray.append(num)
                backtrackingDFS(num, newSet, newArray)
                newArray.popLast()
                backtrackingDFS(num, newSet, newArray)
            }
        }
        
    }

    func permute(_ nums: [Int]) -> [[Int]] {
        self.nums = nums
        for num in nums {
            let set = Set<Int>()
            backtrackingDFS(num, set, [num])
        }
        return result
    }
}

File: 46. Permutations.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        def backTrackingDFS(used: set,perm: List[int]):

            if len(perm) == len(nums):
                result.append(perm[:])
                return

            for num in nums:
                if num not in used:
                    used.add(num)
                    perm.append(num)
                    backTrackingDFS(used,perm)
                    perm.pop()
                    used.remove(num)

        backTrackingDFS(set(),[])
        return result
